### PR TITLE
Pbx no apache [WIP]

### PR DESCRIPTION
**Pull request by jvonau**
### Description of changes proposed in this pull request:
drop apache
### Smoke-tested on which OS or OS's:
lightly tested ubuntu-20.10

### CODE REVIEW

**Comment by holta** [comment]
Wow

**Comment by holta** [comment]
@lemueldsouza can you review this quick?

Test installing it onto a fresh install of RaspiOS if you have time?  As follows:

```
curl d.iiab.io/install.txt | sudo -s 2862
```

Many thanks if you can assist and/or make suggestions!

Refs:

- #1493
- #2762
- PR #2850
- #2856
- PR #2861

File: roles/pbx/tasks/freepbx.yml [hunk 1]
```diff
@@ -124,14 +124,14 @@
     state: file
     mode: "{{ item.mode }}"
   with_items:
-  - { path: '/etc/freepbx.conf', mode: 0664 }
-  - { path: '/var/log/asterisk/freepbx.log', mode: 0666 }
+  - { path: '/etc/freepbx.conf', mode: u=rw,g=r,o=r }
```
Comment (holta) [inline roles/pbx/tasks/freepbx.yml:127]: Is this `/etc/freepbx.conf` change from 664 to 644 on purpose?

(I don't know the official ownership / permissions recommendations!)

File: roles/pbx/tasks/freepbx.yml [hunk 2]
```diff
@@ -118,30 +118,47 @@
...
+  - { path: '/var/log/asterisk/freepbx.log', mode: u=rw,g=rw,o=rw }
+
+- name: FreePBX - fix some permissions for nginx
+  file:
+    path: "{{ item }}"
+    state: directory
+    mode: u=rwx,g=rwx,o=rwx
+  with_items:
+  - /var/www/html/freepbx/admin/assets/less/cache
```
Comment (holta) [inline roles/pbx/tasks/freepbx.yml:136]: Softcode left side of path?  If so using:

```
freepbx_install_dir: /var/www/html/freepbx

File: roles/pbx/tasks/freepbx.yml (hunk 1 above)
Comment (jvonau) [reply roles/pbx/tasks/freepbx.yml:127]: Not sure if official nginx docs exist for freepbx.. I'll explain in greater detail.

File: roles/pbx/tasks/freepbx.yml (hunk 2 above)
Comment (jvonau) [reply roles/pbx/tasks/freepbx.yml:136]: next round of commits

**Comment by jvonau** [comment]
Used https://community.freepbx.org/t/nginx-for-freepbx/61671 as a base for the template, but note that both nginx and php-fpm are running as 'asterisk' in that layout, it's the path of least resistance. The nginx template got to the admin page but with errors about accessing /etc/freepbx.conf 
>ubuntu\@box:/opt/iiab/iiab$ ls -l /etc/freepbx.conf
-rw-rw---- 1 asterisk asterisk 331 Jul 14 06:57 /etc/freepbx.conf

The above needs to be accessible by www-data as that is the user nginx is running as. 
```
Jul 14 19:14:39 box.lan systemd[1]: Starting FreePBX VoIP Server...
Jul 14 19:14:40 box.lan fwconsole[321949]: Running FreePBX startup...
Jul 14 19:14:40 box.lan fwconsole[321949]: Setting Permissions...
Jul 14 19:14:46 box.lan fwconsole[321949]: Setting base permissions...Done in 6 seconds
Jul 14 19:14:46 box.lan fwconsole[321949]: Setting specific permissions...
Jul 14 19:14:46 box.lan fwconsole[321949]:     0 [>---------------------------]
Jul 14 19:14:46 box.lan fwconsole[321949]: Finished setting permissions
Jul 14 19:14:46 box.lan fwconsole[321949]: Starting Asterisk...
Jul 14 19:14:47 box.lan fwconsole[321949]: [>---------------------------] < 1 sec
Jul 14 19:14:47 box.lan fwconsole[321949]: [->--------------------------] 1 sec
Jul 14 19:14:47 box.lan fwconsole[321949]: [-->-------------------------] 1 sec
Jul 14 19:14:48 box.lan fwconsole[321949]: [--->------------------------] 1 sec
Jul 14 19:14:48 box.lan fwconsole[321949]: [---->-----------------------] 2 secs
Jul 14 19:14:49 box.lan fwconsole[321949]: [----->----------------------] 2 secs
Jul 14 19:14:49 box.lan fwconsole[321949]: [------>---------------------] 3 secs
Jul 14 19:14:49 box.lan fwconsole[321949]: [------->--------------------] 3 secs
Jul 14 19:14:50 box.lan fwconsole[321949]: [-------->-------------------] 3 secs
Jul 14 19:14:51 box.lan fwconsole[321949]: [--------->------------------] 4 secs
Jul 14 19:14:51 box.lan fwconsole[321949]: [---------->-----------------] 4 secs
Jul 14 19:14:51 box.lan fwconsole[321949]: [----------->----------------] 4 secs
Jul 14 19:14:51 box.lan fwconsole[321949]: [------------>---------------] 5 secs
Jul 14 19:14:51 box.lan fwconsole[321949]: [------------->--------------] 5 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [-------------->-------------] 5 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [--------------->------------] 6 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [---------------->-----------] 6 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [----------------->----------] 6 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [------------------>---------] 6 secs
Jul 14 19:14:52 box.lan fwconsole[321949]: [------------------->--------] 6 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [-------------------->-------] 6 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [--------------------->------] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [---------------------->-----] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [----------------------->----] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [------------------------>---] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [------------------------->--] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [-------------------------->-] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [--------------------------->] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: [>---------------------------] 7 secs
Jul 14 19:14:53 box.lan fwconsole[321949]: Asterisk Started
Jul 14 19:14:53 box.lan fwconsole[321949]: Running Asterisk post from Core module
Jul 14 19:14:53 box.lan runuser[322174]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:54 box.lan runuser[322174]: pam_unix(runuser:session): session closed for user asterisk
Jul 14 19:14:54 box.lan fwconsole[321949]: Starting Core FastAGI Server...
Jul 14 19:14:54 box.lan runuser[322185]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:55 box.lan runuser[322185]: pam_unix(runuser:session): session closed for user asterisk
Jul 14 19:14:55 box.lan runuser[322196]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:55 box.lan runuser[322196]: pam_unix(runuser:session): session closed for user asterisk
Jul 14 19:14:56 box.lan runuser[322207]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:56 box.lan runuser[322207]: pam_unix(runuser:session): session closed for user asterisk
Jul 14 19:14:57 box.lan runuser[322225]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:57 box.lan runuser[322225]: pam_unix(runuser:session): session closed for user asterisk
Jul 14 19:14:57 box.lan runuser[322240]: pam_unix(runuser:session): session opened for user asterisk by (uid=0)
Jul 14 19:14:58 box.lan runuser[322240]: pam_unix(runuser:session): session closed for user asterisk
```
Note the enforced permissions done by fwconsole, the ansible file mods are undone.... The noted files' permissions were hand changed and the site started to respond correctly but does not persist upon restarting freepbx.

Just need to have /etc/freepbx.conf readable by www-data, the logging write-able and the directories changed was enough to get the site starting to work. Not sure what to do try next add user www-data to the asterisk group or just use PostStart= in the unit file with the same permissions changes that was uncovered. Lets see if we can get a plan of attack together.

**Comment by jvonau** [comment]
> @m-anish writes:
> 
> > My approach would be to try some tested recipe for nginx+pbx, and [then] convert into a playbook. It seems as if jerry is making good progress. If there is somewhere specific he is stuck at, let me know. I am unable right now to be able to test this as dont have a stable working station (am moving between places). But this would be great to have up and running, so people like Gurmet may also benefit from it.

@m-anish see above

**Comment by lemueldsouza** [comment]
Sorry @holta @jvonau, I was away for last few days. Let me try read all the comments I missed and see if I can get it to work.

**Comment by holta** [comment]
> Sorry @holta @jvonau, I was away for last few days. Let me try read all the comments I missed and see if I can get it to work.

Also call me (holt @ unleashkids.org) and @jvonau anytime if that helps us all getting organized here.

_Thanks @lemueldsouza !_

**Comment by jvonau** [comment]
Not testable as is..

**Comment by lemueldsouza** [comment]
Taking cue from @m-anish 's suggestion, I tried to get PBX to work with NGINX as I couldn't find an available recipe

## Details
Raspberry Pi 3B+ with latest raspbian 10
freepbx 15.x, asterisk 16.x, nginx

** I followed the steps from these two links (details in our email thread): **
https://computingforgeeks.com/how-to-install-asterisk-16-with-freepbx-15-on-ubuntu-debian/
and
https://community.freepbx.org/t/nginx-for-freepbx/61671

After installation, I was able to create two chan_pjsip extensions and was successful in make a call across softphones on my cell phones.

Anything you'd like me to try next?
I was just reading about running freepbx as www-data and it seems like people have been having issues as 'asterisk' is apparently being used in the code. Someone also mentions about disabling/uninstalling SIPSTATION module, as that module apparently has this issue.
See https://community.freepbx.org/t/running-freepbx-as-www-data/30017

Looking at this, and considering the fact that we don't know who is using which modules of freepbx with IIAB, should we still go ahead and run as www-data?

@jvonau @holta your thoughts or next steps?

**Comment by holta** [comment]
@lemueldsouza what's your own recommendation?

Should @jvonau rework this PR in any particular directions?

And/or do you want to do that?

**Comment by lemueldsouza** [comment]
> @lemueldsouza what's your own recommendation?

Not sure if it makes sense, but for the upcoming release I would let things run the way it has been before (as 'asterisk') and probably have another experimental version/patch after that to use www-data (so users have an option to chose what they want or fallback if something were to go wrong). 


> And/or do you want to do that?

@jvonau seems to know a lot, he would be the best person to work on it, I could help with testing

**Comment by jvonau** [comment]
The basic pbx layout looks like a container with everything running as asterisk, the goal is to run the nginx webserver as the normal webserver user www-data not as *asterisk* as shown in every bloody howto floating around flogged by 'bloggers' and not software engineers. Apache's use of mpm_itk_module in the pbx role allows apache to run as www-data while accessing the backend DB as asterisk. Think installing asterisk to run as www-data in place of asterisk or some combo of www-data user in asterisk's group and/or asterisk user in www-data group might work but is really low on things I want to debug given the clean room testing needed to figure out what needs attention to be able to drop apache. Upstream alludes to being able to tune the chown command https://wiki.freepbx.org/display/FOP/FreePBX+Chown+Conf so there might be a way around this BS.

In the short term with reports the apache based backend working perhaps a nginx conf file that does a proxy or redirect from '/freebpx' to the alternate apache port would be welcome I think.

**Comment by holta** [comment]
This is Huge Progress having these insights over the past 24h from @lemueldsouza and @jvonau.

Any prototype should be considered, no matter how imperfect (even if testing does not cover the multitude of use cases that can be implemented by Asterisk + FreePBX!) Even if initially perhaps focused on just 1 OS?

I can help with community documentation if so, to make sure grassroots implementers around the world take this very seriously — and assist us in evolving this in whatever useful directions prove most valuable.

> In the short term with reports the apache based backend working perhaps a nginx conf file that does a proxy or redirect from '/freebpx' to the alternate apache port would be welcome I think.

@lemueldsouza is the above [NGINX](https://github.com/iiab/iiab/tree/master/roles/nginx#transition-to-nginx) configuration file something you can possibly help with?  Ask any questions if so — so others can assist you!

**Comment by holta** [comment]
A Giant Thank You to @lemueldsouza who is actively researching options here over the coming week!

Towards more conclusive recommendations, possibly as early as mid-week?

Perhaps in coordination with PR #2850, which can be installed on a fresh OS (with IIAB) as follows:

```
curl d.iiab.io/install.txt | sudo bash -s 2850
```

**Comment by holta** [comment]
Rebased, as a result of: (cosmetic merge conflict removed)

- PR #2879

**Comment by jvonau** [comment]
you are on your own.

**Comment by jvonau** [comment]
Notes for my question 'how do we stop fwconsole from altering some files during chmod' and the answer is to populate freepbx_chown.conf as needed as already done in [Prevent /etc/asterisk/freepbx_chown.conf take over](https://github.com/iiab/iiab/blob/master/roles/pbx/tasks/freepbx.yml#L84-L93)
>ubuntu@box:~$ cat /etc/asterisk/freepbx_chown.conf
; BEGIN ANSIBLE MANAGED BLOCK
[blacklist]
directory = /var/lib/php/sessions
; END ANSIBLE MANAGED BLOCK

options available https://wiki.freepbx.org/display/FOP/FreePBX+Chown+Conf

**Pull request closed without merging**